File: 백준/Gold/2230. 수 고르기/수 고르기.py
import sys

input = sys.stdin.readline
n, m = map(int, input().split())
arr = [int(input()) for _ in range(n)]

if n == 1:  # 예외 처리 (n이 1이면 차이를 만들 수 없음)
    print(0)
    sys.exit()

arr.sort()

left = 0
right = 1
answer = sys.maxsize

while right < n:
    gap = arr[right] - arr[left]

    if gap >= m:
        answer = min(answer, gap)
        left += 1
        if left == right:  # left와 right가 같아지면 right도 증가
            right += 1
    else:
        right += 1

print(answer)

File: 백준/Gold/2230. 수 고르기/수 고르기.java
import java.util.*;
import java.io.*;
public class Main{
    static int N,M;
    static Integer[] arr;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new Integer[N];
        for (int i=0; i<N; i++){
            arr[i] = Integer.parseInt(br.readLine());
        }
        int ans = Integer.MAX_VALUE;
        Arrays.sort(arr);
        for (int i=0; i<N-1; i++){
            ans = Math.min(ans, binarySearch(i));
        }
        System.out.println(ans);
    }

    static int binarySearch(int standard){
        int left = standard+1;
        int right = N-1;
        int mid, target;
        int diff = Integer.MAX_VALUE;
        while (left<=right){
            mid = left + (right-left)/2;
            target = arr[mid] - arr[standard];
            if (target < M){
                left = mid+1;
            } else {
                right = mid-1;
                diff = target;
            }

        }
        return diff;
    }
}